### PR TITLE
feat/helmChart: initial support for installing the operator using helm charts

### PR DESCRIPTION
**Pull request by nujragan93**
Hey there 👋

This PR adds initial support for helm chart
 https://github.com/dragonflydb/dragonfly-operator/issues/168

### CODE REVIEW

File: charts/dragonfly-operator/templates/namespace.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/templates/namespace.yaml]: I don't think we should be including a namespace resource, helm lets you install into namespaces via 

```helm install --create-namespace --namespace dragonfly ...```

and that will set all the resources to that namespace in the templates.

Safe to also say I do not see including the namespace resource being done in other helm charts as well.

**Comment by onedr0p** [comment]
I don't see any workflow changes to publish the Helm chart (e.g. https://github.com/dragonflydb/dragonfly/blob/b994f934d3311925d4dc48dcb87f5c560d478b27/.github/workflows/docker-release.yml#L82-L96)

File: charts/dragonfly-operator/values.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/values.yaml]:
```suggestion
  resources: {}
```

Resources should not have a default

File: charts/dragonfly-operator/values.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/values.yaml]:
```suggestion
  affinity: {}
```

I don't see why there needs to be a default set for the affinity. Better to leave this unset and leave it up to the user.

File: charts/dragonfly-operator/values.yaml
Comment (nujragan93) [reply charts/dragonfly-operator/values.yaml]: missed commenting out those! thanks

File: charts/dragonfly-operator/values.yaml
Comment (nujragan93) [reply charts/dragonfly-operator/values.yaml]: missed commenting out those! thanks

**Comment by nujragan93** [comment]
> Thanks for tackling this @nujragan93 I don't see any workflow changes to publish the Helm chart (e.g. what is done in https://github.com/dragonflydb/dragonfly/blob/main/.github/workflows/docker-release.yml)

that seems like a OCI repo for dragonflyDb, does the operator need to have a separate repo ? do we need to add tests to the CI before pushing ?

**Comment by onedr0p** [comment]
> that seems like a OCI repo for dragonflyDb, does the operator need to have a separate repo ?

They seem to be packaging up the Helm chart into a [OCI artifact](https://helm.sh/blog/storing-charts-in-oci/) which seems like that could be done here as well and published under https://github.com/orgs/dragonflydb/packages?repo_name=dragonfly-operator

> do we need to add tests to the CI before pushing ?

It looks like they have some CI tests in https://github.com/dragonflydb/dragonfly/tree/main/contrib/charts/dragonfly/ci so probably might be a good idea to have some here as well.

File: charts/dragonfly-operator/values.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/values.yaml]:
```suggestion
    tag: ""
```

This should come from the Chart.yaml `appVersion` field as implemented below

File: charts/dragonfly-operator/Chart.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/Chart.yaml]:
```suggestion
appVersion: "v1.1.2"
```

This is the current operator version

File: charts/dragonfly-operator/templates/deployment.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/templates/deployment.yaml]:
```suggestion
          image: "{{ .Values.manager.image.repository }}:{{ .Values.manager.image.tag | default .Chart.AppVersion  }}"
```

Should default to the Chart.yaml `appVersion`

File: charts/dragonfly-operator/values.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/values.yaml:65]:
```suggestion
    repository: ghcr.io/dragonflydb/operator
```

Comment (onedr0p) [reply charts/dragonfly-operator/values.yaml:65]: Not sure if GHCR is better since it is published here?

File: charts/dragonfly-operator/templates/crds.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/templates/crds.yaml:1]: There's a long lengthy debate on how CRDs should be managed in Helm charts. Either by what you did here, or dropping them un-templated in a `crds` folder and letting helm install them by default. If the dragonfly team is okay with this method then it shouldn't be a problem moving forward.

I honestly prefer the `crds` folder method they describe [here](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you) but there are some caveats. The main one being the CRDs will not be upgraded even if you change them here in the chart. There are ways around that... Flux/Argo natively supports upgrading the chart as well as the CRDs while other CI/CD tools might have to `kubectl apply` the CRDs while upgrading the Chart.

File: charts/dragonfly-operator/values.yaml
Comment (nujragan93) [reply charts/dragonfly-operator/values.yaml:65]: @Abhra303 @Pothulapati any ideas?

**Comment by Pothulapati** [comment]
Hey @nujragan93, Thanks for working on this.

+1 to both of these questions. We can have similar CI linting, generation tests like Dragonfly. Feel free to add a simpler verison of that. 

Yep, We should be using Github Packages to store and publish the Helm Chart.

(Reviewing the Helm Chart itself meanwhile)

File: charts/dragonfly-operator/values.yaml
Comment (Pothulapati) [reply charts/dragonfly-operator/values.yaml:65]: We should use docker.dragonflydb.io. It is just a gateway to ghcr, so its all the same.

**Comment by nujragan93** [comment]
> Hey @nujragan93, Thanks for working on this.
> 
> +1 to both of these questions. We can have similar CI linting, generation tests like Dragonfly. Feel free to add a simpler verison of that.
> 
> Yep, We should be using Github Packages to store and publish the Helm Chart.
> 
> (Reviewing the Helm Chart itself meanwhile)

I have made some workflow changes to push helm chart to github(dont know how to test it)
This PR seems already very big, I feel the ci linting and tests should be made a separate PR.

**Comment by onedr0p** [comment]
@nujragan93 pipelines looks good. I also suggest we separate the PRs for adding ci linting and tests after this is merged.

**Comment by nujragan93** [comment]
> @nujragan93 pipeline looks good. I agree we should separate the PRs for adding ci linting and tests after this is merged.

agreed.

File: charts/dragonfly-operator/templates/deployment.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/templates/deployment.yaml:38]: It would be nice to allow for extra args here.

File: charts/dragonfly-operator/templates/deployment.yaml
Comment (onedr0p) [inline charts/dragonfly-operator/templates/deployment.yaml:60]: It would be nice to allow for extra args here.

**Comment by onedr0p** [comment]
@Pothulapati any update here?

**Comment by onedr0p** [comment]
@Pothulapati, or anybody?

File: .github/workflows/ci.yml
Comment (Pothulapati) [inline .github/workflows/ci.yml:87]: nit: Is there a need to commit the version into repository? considering we publish the chart as an artifact? :thinking:

File: charts/dragonfly-operator/templates/tests/test-connection.yaml
Comment (Pothulapati) [inline charts/dragonfly-operator/templates/tests/test-connection.yaml:1]: Not really sure what this file is for :thinking:

**Review by Pothulapati — COMMENTED** [review]
Added some nits and questions! Code looks mostly good!

Considering Helm is very errror prone (easy to add formatting errors), We should be adding linting tests like https://github.com/dragonflydb/dragonfly/blob/main/.github/actions/lint-test-chart/action.yml, Otherwise we have no way to make sure these charts actually work. This is very important considering we have so many values, etc.

We can merge it once we see that these tests pass.

Thanks! Really sorry for missing the review on this.

File: charts/dragonfly-operator/templates/tests/test-connection.yaml
Comment (onedr0p) [reply charts/dragonfly-operator/templates/tests/test-connection.yaml:1]: https://helm.sh/docs/topics/chart_tests/

File: .github/workflows/ci.yml
Comment (onedr0p) [reply .github/workflows/ci.yml:87]: It's also being done [here](https://github.com/dragonflydb/dragonfly/blob/6b8a8b6410c84f0d03444a6700f54fb806270253/.github/workflows/docker-release.yml#L93) in dragonflydb/dragonfly, so points for consistency? 😆

**Comment by onedr0p** [comment]
Anything I can do here to help move this along @Pothulapati @nujragan93 ?

File: .github/workflows/ci.yml
Comment (Pothulapati) [reply .github/workflows/ci.yml:87]: Sure!

**Comment by Pothulapati** [comment]
@onedr0p Feel free to  create a new branch from this branch and add your commits with the changes that I requested i.e test the chart. I'm happy to follow up with a review.

**Comment by Pothulapati** [comment]
>  @Pothulapati you are a maintainer of this repo which gives you the ability to edit this PR with those said changes. 

Yes, but that does not mean I have enough bandwidth to work on everything :/

> If that's not something you want to do I'll go ahead and create a branch off this and open a PR with your requested changes.

If you are interested, and want to help, Feel free!

**Comment by onedr0p** [comment]
@Pothulapati  After looking this over... there's quite a bit of CI changes that need to be done to support this. Why is it not possible to merge this PR and iterate on the changes in future updates?

**Comment by nujragan93** [comment]
> @Pothulapati After looking this over... there's quite a bit of CI/GitHub workflow changes that need to be done to support what you ask because the workflow you have in this repo is only for releases / pushing to main. There is no testing workflow to expand upon like here https://github.com/dragonflydb/dragonfly/tree/main/.github/workflows
> 
> Why is it not possible to merge this PR and iterate on the changes in future updates? As @nujragan93 already mentioned this PR is pretty big already.

I agree, if the workflows are setup, then I would jump in and iterate on tests with the helm charts.

**Review by Abhra303 — APPROVED** [review]

**Comment by Pothulapati** [comment]
Sorry folks but this is merged now! Looking forward to the next iterations!

**Comment by FlorisFeddema** [comment]
Does this need a new operator release first? Seems like the package is not yet available in the ghcr

**Comment by Pothulapati** [comment]
Just FYI, we had to remove the release changes as it wasn't working https://github.com/dragonflydb/dragonfly-operator/pull/201

**Comment by alexanderccc** [comment]
@Pothulapati do you need the commit back into the main branch? Assuming it's failing there, I think this might be fixed in 2 ways

Keep the chart and app separate, and have another workflow for helm, this also means having separate versions for the chart which might increase the complexity a bit as there would be 2 different artifacts that need to be maintained and released as part of the same repo with different versions.

Or keep it with matching versions and consider helm as part of the app. Any changes to helm are released with a new version of the app as well (common helm headache) and keep the workflow with sed for version change, login, package and push. For local install (from the cloned repo) there could be a note in the readme stating that the chart and app version need to be updated.

If Helm gets out of hand (which it might), this can be separated into its own repo where you only have the chart.

**Comment by Aym3nTN** [comment]
The fix is easy. Simply add`fetch-depth: 0` to `actions/checkout@v3` step and `git push` in the step where you update Chart.yaml file with the tag version. 
But the concern here is that the Helm Chart version and Operator version are different, the former is at v1.x.x > and the former at v0.1.x. Not the same, I'd suggest working on improving the version management of both components. I'm opening another issue for this initiative.

**Pull request merged**
